### DISCUSSION TITLE
subflow

### DISCUSSION THREAD

**Discussion by phoddie** [phoddie/node-red-mcu#38] [Ideas]
I've been looking into implementing [subflows](https://nodered.org/docs/user-guide/editor/workspace/subflows) in Node-RED MCU Edition. It is not a small project. Subflows bundle together a set of nodes for reuse within a project and publishing to npm to share. Subflows are perhaps the largest unimplemented structucal feature in Node-RED MCU Edition. It looks possible to support them, but there are many details to consider. I'm opening this discussion thread to share my current thoughts and notes. 

A subflow is node that contains a collection of nodes. A subflow has these characteristics:

- 0 or 1 inputs
- 0 or more outputs
- A subflow may be instanced one or more times in other flows
- Uncaught errors are delivered as errors to subflow instance node
- Subflows may contain a special status node with a single input. Messages delivered to this special status are set as the status of the subflow instance node
- Subflows may contain environment variables
- Subflows change the behavior of some automatic environment variables, including `NR_FLOW_ID` and `NR_NODE_PATH`.
- A subflow may contain other subflows

An instance of a subflow behaves like an ordinary node. In the Node-RED Editor, a subflow looks like an ordinary node. But the implementation is a set of Node-RED nodes rather than JavaScript code.

Each instance of a subflow creates a clone of the subflow and the nodes it contains. The cloned subflow and nodes all received unique `ID` values. This requires updating all connections within the subflow (`wire`, `scope`, `link`, etc) the new `ID` values. The step of assigning new `ID` values for each instance ensures that each node in the Node-RED runtime has a unique ID, even when a subflow is instanced more than once.

Most of the work to support subflows can be done at build time in the `nodered2mcu` tool. Each instance of a subflow creates a new flow instance. The subflow template, including all its child nodes, should not appear in the output.

In theory, a subflow with no environment variables could be instanced with only nodes (though this would change the value of `NR_NODE_PATH`). This optimization doesn't seem worth pursuing because it looks complicated and environment variables seem essential to configuring subflows.

The Node-RED Editor renders the the inputs and outputs of a subflow as  special nodes. However, they are not nodes, but just wires. The `nodered2mcu` tool needs to connect the inputs and outputs of each subflow instance node to the corresponding nodes in their containing flow.

**Comment by ralphwetzel** [comment]
> The subflow template, including all its child nodes, should not appear in the output.

At least the id of the subflow "node" needs to be kept in the output - to be able to assign error & status messages to it's visual representation in the editor.

**Reply by phoddie** [reply]
Agreed, that's needed. What should not be output is the template which consists of  the JSON record of `type` `"subflow"` and all nodes with a `z` value equal to the `id` (for example `"98a1ae9f57c26e96 "`) of the subflow. The subflow instance node has a `type` `"subflow:98a1ae9f57c26e96" and the `z` value of its containing flow.
